src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs

```diff
@@ -0,0 +1,88 @@
+    internal class UseExplicitTypingCodeFixProvider : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds =>
+            ImmutableArray.Create(IDEDiagnosticIds.UseExplicitTypingDiagnosticId);
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+            var root = await document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+            var token = root.FindToken(span.Start);
+            if (!token.Span.IntersectsWith(span))
+            {
+                return;
+            }
+
+            var node = root.FindNode(span, getInnermostNodeForTie: true);
+
+            var codeAction = new MyCodeAction(
+                CSharpFeaturesResources.UseExplicitType,
+                c => HandleDeclaration(document, root, node, context.CancellationToken));
+
+            context.RegisterCodeFix(codeAction, context.Diagnostics.First());
+        }
+
+        private async Task<Document> HandleDeclaration(Document document, SyntaxNode root, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var declarationContext = node.Parent;
+
+            TypeSyntax typeSyntax;
+            if (declarationContext is VariableDeclarationSyntax)
+            {
+                typeSyntax = ((VariableDeclarationSyntax)declarationContext).Type;
+            }
+            else if (declarationContext is ForEachStatementSyntax)
+            {
+                typeSyntax = ((ForEachStatementSyntax)declarationContext).Type;
+            }
+            else
+            {
+                Debug.Assert(false, $"unhandled kind {declarationContext.Kind().ToString()}");
+                return document;
+            }
+
+            var typeSymbol = semanticModel.GetTypeInfo(typeSyntax).ConvertedType;
+
+            var typeName = SyntaxFactory.ParseTypeName(typeSymbol.ToDisplayString())
+                                            .WithLeadingTrivia(node.GetLeadingTrivia())
+                                            .WithTrailingTrivia(node.GetTrailingTrivia())
+                                            .WithAdditionalAnnotations(Simplifier.Annotation);
+
+            Debug.Assert(!typeName.ContainsDiagnostics, "Explicit type replacement likely introduced an error in code");
+
+            var newRoot = root.ReplaceNode(node, typeName);
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(string title, Func<CancellationToken, Task<Document>> createChangedDocument) :
+                base(title, createChangedDocument)
```

Comment:
You need to pass 'title' along for the equivalence key.

Also, did you implement the batch fix provider?  we'll want that for this feature.